File: src/eventCodes.js
// Major/minor event codes, per the ISAPI event category tables.
const MAJOR = {
  1: 'alarm',
  2: 'exception',
  3: 'operation',
  5: 'event',
};

// Minor codes under MAJOR_EVENT (5). The minor code — not currentVerifyMode —
// is what authoritatively says which credential was used.
const MINOR = {
  27: { name: 'exitButtonPressed', method: 'button', success: true },
  38: { name: 'cardAuthSuccess', method: 'card', success: true },
  75: { name: 'faceAuthSuccess', method: 'face', success: true },
  76: { name: 'faceAuthFail', method: 'face', success: false },
  113: { name: 'fingerprintAuthSuccess', method: 'fingerprint', success: true },
};

function describe(major, minor) {
  const entry = major === 5 ? MINOR[minor] : undefined;
  return {
    majorName: MAJOR[major] ?? null,
    minorName: entry?.name ?? null,
    method: entry?.method ?? null,
    success: entry?.success ?? null,
  };
}

module.exports = { MAJOR, MINOR, describe };

File: isup-server/src/main/java/com/hrm/isup/event/EventCodes.java
package com.hrm.isup.event;

import com.hrm.isup.model.AccessEvent;

import java.util.Map;

/**
 * Canonical Hikvision access-event codes (major 5). The MINOR code — not
 * currentVerifyMode — authoritatively names the credential that matched.
 *
 * Ported from the Node reference (src/eventCodes.js / lib/hikvision-event-parser).
 * Centralized so both the inbound webhook gateway and (later) EventPollService
 * classify events the same way.
 */
public final class EventCodes {

    public static final int MAJOR_EVENT = 5;

    /** A resolved description of a minor code. */
    public record Desc(String name, String method, Boolean success) {}

    private static final Map<Integer, Desc> MINOR = Map.of(
            27,  new Desc("exitButtonPressed",      "button",      true),
            38,  new Desc("cardAuthSuccess",        "card",        true),
            75,  new Desc("faceAuthSuccess",        "face",        true),
            76,  new Desc("faceAuthFail",           "face",        false),
            113, new Desc("fingerprintAuthSuccess", "fingerprint", true)
            // Add device-specific minors here as they are observed (see event.raw).
    );

    private EventCodes() {}

    /** name/method/success for a major+minor, or all-null when unmapped. */
    public static Desc describe(int major, int minor) {
        Desc d = major == MAJOR_EVENT ? MINOR.get(minor) : null;
        return d != null ? d : new Desc(null, null, null);
    }

    /**
     * A keep-alive (no credential) — not a real punch. Matches the Node rule:
     * non-access envelope, or no employeeNo/cardNo and no resolved method.
     */
    public static boolean isHeartbeat(AccessEvent e) {
        if (e == null) return true;
        if (e.eventType != null && !"AccessControllerEvent".equals(e.eventType)) return true;
        if ("exitButtonPressed".equals(e.eventName)) return false; // real, but has no person
        boolean noEmp = e.employeeNo == null || e.employeeNo.isEmpty();
        boolean noCard = e.cardNo == null || e.cardNo.isEmpty();
        return noEmp && noCard && e.verifyMethod == null;
    }
}
